### PR TITLE
prov/rxm,verbs: Simplify XRC connection establishment prototype/experimental 

### PR DESCRIPTION
**Pull request by swelch**
This set of commits is an experimental/prototype for reducing the complexity of the Verbs provider XRC transport connection setup. The XRC transport connects an XRC initiator (send) QP to a remote XRC target (receive) QP, establishing simplex operation. In the current implementation, the Verbs provider hides this simplex nature from RxM by creating the connection in the reverse direction as part of the RxM connect request.

This code removes this hidden reciprocal connection from the Verbs provider, and allows RxM which already knows the address to EP mapping to drive a connection operation in both directions. To do this the notion of a simplex transport based EP is introduced and defined to allow a single EP to both connect and accept a connection (to establish duplex operation). Support of the OFI_OPS_SIMPLEX_CM ops by the core provider indicates this behavior and extends the fi_accept() operation to pass
connection request fi_info to an existing EP.

The advantages of this approach:
- Maintainability, it significantly reduces the amount of XRC setup code while only minimally adding to the RxM code complexity
- Reduces the connect time in that the connection request does not have to delay waiting for the reverse connection to setup
- Maintaining use of a single EP for duplex operation keeps from increasing memory consumption

The commit on the tip of this PR removes the XRC logic that releases RDMA CM IDs after physical connections are established. The value of this commit must be examined at greater scale and determined if appropriate. While it reduces the number of CM
messages exchanged during connection setup, it requires a greater number of RDMA CM IDs exist during steady state operation. There are previous versions of MOFED that have limitations in port space that restricted scale. NOTE: This commit may be removed independent of the others to compare operation while still releasing RDMA CM ID resources.

### CODE REVIEW

**Comment by shefty** [comment]
Adding do not merge label until after v1.12 release

**Comment by swelch** [comment]
> Adding do not merge label until after v1.12 release

Thanks!

**Comment by wzamazon** [comment]
AWS failure is due to a recent change in CI script. We are working on resolving it.

**Comment by wzamazon** [comment]
bot:aws:retest

**Comment by wzamazon** [comment]
The CI script issue was fixed. the new AWS CI failure is because ubuntu software repository temporarily unavailable.

**Comment by wzamazon** [comment]
bot:aws:retest

**Comment by swelch** [comment]
> The CI script issue was fixed. the new AWS CI failure is because ubuntu software repository temporarily unavailable.

Thanks for the update and kicking it off again.

File: prov/rxm/src/rxm_cq.c [hunk 1]
```diff
@@ -480,10 +487,15 @@ static ssize_t rxm_rndv_xfer(struct rxm_ep *rxm_ep, struct fid_ep *msg_ep,
 		if (ret)
 			return ret;
 		total_len -= copy_len;
-		ret = rxm_ep->rndv_ops->xfer(msg_ep, iov, desc, count, 0,
-			       remote_hdr->iov[i].addr, remote_hdr->iov[i].key,
-			       context);
 
+		if (conn_sts)
+			ret = conn_sts;
+		else
```
Comment (shefty) [inline prov/rxm/src/rxm_cq.c:493]: I'm lost on the purpose of this check.  Why did we continue to here?  I think conn_sts (not sure what sts means btw) is -FI_EAGAIN.  Splitting the check of conn_sts so far from where it was set is confusing.

File: prov/rxm/src/rxm_cq.c [hunk 2]
```diff
@@ -964,8 +975,10 @@ ssize_t rxm_rndv_send_wr_data(struct rxm_rx_buf *rx_buf)
 			  rx_buf->recv_entry->rxm_iov.iov,
 			  rx_buf->recv_entry->rxm_iov.count, rx_buf->mr);
 
-	ret = fi_send(rx_buf->conn->msg_ep, &buf->pkt, sizeof(buf->pkt) +
-		      sizeof(struct rxm_rndv_hdr), buf->hdr.desc, 0, rx_buf);
+	ret = rxm_check_duplex_conn(&rx_buf->conn->handle);
```
Comment (shefty) [inline prov/rxm/src/rxm_cq.c:978]: Similar question as above.  If the connection isn't ready, I would have expected a failure sooner than this call.

File: prov/rxm/src/rxm_cq.c [hunk 3]
```diff
@@ -1048,6 +1061,10 @@ static ssize_t rxm_atomic_send_resp(struct rxm_ep *rxm_ep,
 	atomic_hdr->status = htonl(status);
 	atomic_hdr->result_len = htonl(result_len);
 
+	ret = rxm_check_duplex_conn(&rx_buf->conn->handle);
```
Comment (shefty) [inline prov/rxm/src/rxm_cq.c:1064]: here as well

File: prov/verbs/src/fi_verbs.h [hunk 4]
```diff
@@ -365,8 +343,7 @@ struct vrb_domain {
 
 	ssize_t		(*send_credits)(struct fid_ep *ep, uint64_t credits);
 
-	/* Indicates that MSG endpoints should use the XRC transport.
-	 * TODO: Move selection of XRC/RC to endpoint info from domain */
+	/* Indicates that MSG endpoints should use the XRC transport. */
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.h:346]: This is separate from this change, but if flags is a bool, we should convert it to that, and rename it to indicate what it represents (use_xrc), so we can drop this comment.  If flags is not a bool, then this comment should still go away and be placed where the flag is defined, if that flag name doesn't make it obvious.

File: prov/rxm/src/rxm_cq.c [hunk 5]
```diff
@@ -468,6 +468,13 @@ static ssize_t rxm_rndv_xfer(struct rxm_ep *rxm_ep, struct fid_ep *msg_ep,
 	struct iovec iov[RXM_IOV_LIMIT];
 	void *desc[RXM_IOV_LIMIT];
 	ssize_t ret = FI_SUCCESS;
+	ssize_t conn_sts = FI_SUCCESS;;
+
+	if (rxm_ep->rndv_ops == &rxm_rndv_ops_read) {
+		conn_sts = rxm_check_duplex_conn(&conn->handle);
+		if (conn_sts && conn_sts != -FI_EAGAIN)
+			return conn_sts;
+	}
```
[the remark this answers is not in the document]
Comment (swelch) [reply prov/rxm/src/rxm_cq.c:477]: > It seems that the check for a usable connection should have occurred sooner than this call.

The intent of delaying the check for the TX connection back to the originator of a request was to simplify the initialization of a deferred entry should the connection in that direction not yet exist. Where the check is now we can easily initialize the defer structure and just let the existing code do that initialization and enqueue it, in addition at this point the RX buffer is no longer needed. Another consideration was to avoid introducing the overhead of the check until it was in code that actually needed to send something back to originator of the software protocol request. With that said, I had debated putting the check earlier in the RX CQ handling processing and it can be done. It would need to be deep enough into the CQ path so that only protocols that require sending something back to the requestor make the check, as otherwise we would create a connection that is not required (i.e. if only RMA and eager processing we may not require a connection back to the requestor depending on application usage). If moved I would want to create a new defer type and hang on to the RX buffer until the deferred entry could be processed, allowing us to re-enter the RX CQ processing. This adds more new code but might make it easier to understand. The new deferred processing would need to first validate the connection (so we don't create duplicate deferred entries for the same request) and then call back into the *right* spot of the CQ processing path if the TX connection has completed. I had taken some comfort in not creating a new code path. Do the extra details alter your opinion or do you want me to investigate moving it up earlier in the CQ processing? Thanks.

File: prov/rxm/src/rxm_conn.c [hunk 6]
```diff
@@ -557,70 +563,126 @@ void rxm_cmap_process_reject(struct rxm_cmap *cmap,
 	}
 }
 
-int rxm_cmap_process_connreq(struct rxm_cmap *cmap, void *addr,
-			     struct rxm_cmap_handle **handle_ret,
-			     uint8_t *reject_reason)
+static int
+rxm_cmap_get_connreq_handle(struct rxm_cmap *cmap, void *addr,
```
[the remark this answers is not in the document]
Comment (swelch) [reply prov/rxm/src/rxm_conn.c:567]: > Rename to rxm_get_cmap_handle() or rxm_cmap_get_handle()? Is this handle only for a connection request?

It is intended to only be used as part of a connection request processing. Basically it is common code that is shared between processing a normal/duplex connection request and a new/simplex connection request.

[the remark this answers is not in the document]
Comment (shefty) [reply prov/rxm/src/rxm_conn.c:567]: You're only using the handle for connection request processing, but the handle is just a handle from the cmap, correct?  That is, the handle is of type rxm_cmap_handle, and all this function does is return that.

[the remark this answers is not in the document]
Comment (swelch) [reply prov/rxm/src/rxm_conn.c:567]: I agree, didn't mean to imply that I was opposed to the name change. I will update the name. Thanks.

File: prov/verbs/src/fi_verbs.h (hunk 4 above)
Comment (swelch) [reply prov/verbs/src/fi_verbs.h:346]: > This is separate from this change, but if flags is a bool, we should convert it to that, and rename it to indicate what it represents (use_xrc), so we can drop this comment. If flags is not a bool, then this comment should still go away and be placed where the flag is defined, if that flag name doesn't make it obvious.

I agree, I will open create a commit for this outside of this PR. Thanks.

File: prov/rxm/src/rxm_conn.c [hunk 7]
```diff
@@ -184,8 +188,7 @@ ssize_t rxm_get_conn(struct rxm_ep *rxm_ep, fi_addr_t addr,
 	assert(rxm_ep->util_ep.tx_cq);
 	handle = rxm_cmap_acquire_handle(rxm_ep->cmap, addr);
 	if (!handle) {
-		ret = rxm_cmap_alloc_handle(rxm_ep->cmap, addr,
-					    RXM_CMAP_IDLE, &handle);
+		ret = rxm_cmap_alloc_handle(rxm_ep->cmap, addr, &handle);
```
[the remark this answers is not in the document]
Comment (swelch) [reply prov/rxm/src/rxm_conn.c:191]: > It would be easier to check the changes if this and similar changes (rxm_cma_init_handle) were separated out and applied before adding the rx_state.

I'll do this and refresh the PR, thanks.

File: prov/rxm/src/rxm_conn.c [hunk 8]
```diff
@@ -557,70 +563,126 @@ void rxm_cmap_process_reject(struct rxm_cmap *cmap,
...
-			goto unlock;
+			ret = rxm_cmap_alloc_handle(cmap, *fi_addr, handle);
+		if (!ret) {
+			if (cmap->simplex)
+				RXM_CM_UPDATE_RX_STATE(*handle,
+						       RXM_CMAP_CONNREQ_RECV);
+			else
+				RXM_CM_UPDATE_STATE(*handle,
+						    RXM_CMAP_CONNREQ_RECV);
+		}
```
[the remark this answers is not in the document]
Comment (swelch) [reply prov/rxm/src/rxm_conn.c:593]: > I find the structure of this function in its original form hard to follow, and I think it's because of the duplicate fi_addr check. Maybe this would be easier?
> 
> ```
> if (fi_addr == FI_ADDR_NOTAVAIL) {
>     handle = rxm_cmap_get_handle_peer(cmap, addr);
>     if (!handle)
>         ret = rxm_cmap_alloc_handle_peer(cmap, addr, handle);
> } else {
>     handle = rxm_cmap_acquire_handle(cmap, fi_addr);
>     if (!handle)
>         ret = rxm_cmap_alloc_handle(cmap, fi_addr, handle);
> }
> ```

I agree, I will add a commit to restructure the original code before applying the commits required by this PR. Thanks.

File: prov/rxm/src/rxm_conn.c [hunk 9]
```diff
@@ -783,10 +857,24 @@ int rxm_cmap_bind_to_av(struct rxm_cmap *cmap, struct util_av *av)
 	return ofi_av_elements_iter(av, rxm_cmap_update_addr, (void *)cmap);
 }
 
+static int rxm_no_simplex_cm_accept(struct fid_ep *ep, struct fi_info *info,
+				    const void *param, size_t paramlen)
+{
+	/* Simplex transport extension not required */
+	return fi_accept(ep, param, paramlen);
```
[the remark this answers is not in the document]
Comment (swelch) [reply prov/rxm/src/rxm_conn.c:864]: > Should we ever be in the path if simplex = false? Should we assert() here?

I think we are good here, the function to open the ops controls both the replacing these default ops and setting the flag indicating simplex cmap operation. However, adding the assert seems like a good idea in that it is a requirement and may detect a later code modification error. I'll add it. Thanks.

File: prov/rxm/src/rxm_conn.c [hunk 10]
```diff
@@ -1108,15 +1213,19 @@ rxm_msg_process_connreq(struct rxm_ep *rxm_ep, struct fi_info *msg_info,
...
+	}
 
 	cm_data.accept.server_conn_id = rxm_conn->handle.key;
 	cm_data.accept.rx_size = rxm_conn_get_rx_size(rxm_ep, msg_info);
 
-	ret = fi_accept(rxm_conn->msg_ep, &cm_data.accept.server_conn_id,
-			sizeof(cm_data.accept));
+	ret = rxm_ep->cmap->simplex_cm_ops->accept(rxm_conn->msg_ep, msg_info,
+						   &cm_data.accept.server_conn_id,
+						   sizeof(cm_data.accept));
```
[the remark this answers is not in the document]
Comment (swelch) [reply prov/rxm/src/rxm_conn.c:1239]: > It's confusing for a duplex connection to call a simplex cm op.

I see your point. Would something like cm_ex_ops be better? Or I could do a comparison of the ops pointer to the default function table pointer or look at cmap->simplex to select calling fi_accept() or this simplex version. It literally is called only from this one spot and this is control path so the extra if is not significant. What is your preference?  Thanks.

[the remark this answers is not in the document]
Comment (shefty) [reply prov/rxm/src/rxm_conn.c:1239]: Of the choices you mentioned, checking simplex to call the right function would seem the most clear.

File: prov/rxm/src/rxm_conn.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (swelch) [reply prov/rxm/src/rxm_conn.c:567]: @shefty  - After looking further I think the name should stay as is. While this routine will return an existing or newly allocated cmap handle, it is meant to be called from connection request processing only and will initialize a newly allocated handles to the connection request received state.

File: prov/rxm/src/rxm_cq.c (hunk 1 above)
Comment (swelch) [reply prov/rxm/src/rxm_cq.c:493]: @shefty - It was just to be more efficient and not repeatedly call the function that checks if connectivity has been established within a loop. I will look at reworking this code to see if I can make it easier to follow (if I decide to move the check for duplex connectivity further up in the processing I will remove it). Thanks.

**Comment by swelch** [comment]
@shefty - I broke out several of the existing code refactoring changes that enable the simplex transport support into separate commits to make the changes easier to follow. I also, removed the changes to make the duplex connectivity optional, RxM will make certain duplex connectivity has been established before permitting TX operations. While this is required for RMA and eager/SAR messaging only, it removes the code that validates duplex connectivity and/or defers RxM protocol responses from the CQ path. This is still experimental.

**Review by shefty — COMMENTED** [review]
reviewed first 5 patches.  The first 4 look ready to separate out and merge.  I think there may be a way to simplify the state tracking.  See my review comment for ideas.

File: prov/rxm/src/rxm_conn.c [hunk 11]
```diff
@@ -613,9 +612,7 @@ int rxm_cmap_process_connreq(struct rxm_cmap *cmap, void *addr,
 		} else {
 			FI_DBG(cmap->av->prov, FI_LOG_EP_CTRL,
 				"Endpoint connects to itself\n");
-			ret = rxm_cmap_alloc_handle_peer(cmap, addr,
-							  RXM_CMAP_CONNREQ_RECV,
-							  &handle);
+			ret = rxm_cmap_alloc_handle_peer(cmap, addr, &handle);
```
[the remark this answers is not in the document]
Comment (swelch) [reply prov/rxm/src/rxm_conn.c:684]: Thanks. I was setting the handle state after the check for the bad return since the handle is not allocated in that case. I was attempting to follow the same logic as the original. Looking back at this now though, the RXM_CM_UPDATE_STATE(handle, RXM_CMAP_CONNREQ_RECV); that follows the error check in the else clause is redundant. That code will fall through to the next case which will set it. I will remove the extra set. Let me know if I addressed the right area in the code or further clarification is needed.

File: prov/rxm/src/rxm.h [hunk 12]
```diff
@@ -164,15 +164,27 @@ extern char *rxm_cm_state_str[];
...
 
+/* Maintain separate RX state for simplex connections */
+#define RXM_CM_UPDATE_RX_STATE(handle, new_state)			\
+	do {								\
+		FI_DBG(&rxm_prov, FI_LOG_EP_CTRL, "[CM] handle: "	\
+		       "%p RX %s -> %s\n", handle,			\
+		       rxm_cm_state_str[(handle)->rx_state],		\
+		       rxm_cm_state_str[new_state]);			\
+		(handle)->rx_state = new_state;				\
+	} while (0)
```
[the remark this answers is not in the document]
Comment (swelch) [reply prov/rxm/src/rxm.h:181]: I like the idea, and think it will help in further simplifying of the code. I'll take a pass at it and refresh this commit. Thanks again.

File: prov/verbs/src/fi_verbs.h (hunk 4 above)
Comment (swelch) [reply prov/verbs/src/fi_verbs.h:346]: @shefty - Looks like the flag field is now used to indicate the domain supports ODP as well. I'm adding a separate change to clean up the comment outside of this PR.

**Comment by swelch** [comment]
@shefty - Pulled a few commits out that I had submitted in different PR, rebased on latest master, added commit maintain rxm handle TX and RX state separately, and updated simplex code to use that. Still just evaluating at this point.

**Comment by jswaro** [comment]
This needs a little more work. It increases the CM time for base MPI connect times when creating a fully connected graph. I'm seeing an increase of ~50% total connection time (~2min without patch to ~3min with patch) at high process density per node in this implementation.

**Comment by shefty** [comment]
Older PR with no recent activity.  Closing.  Please re-open when ready for additional review.
